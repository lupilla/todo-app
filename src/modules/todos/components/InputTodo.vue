<template>
  <div id="newTodo">
    <input
      type="text"
      class="new-todo"
      v-model="name"
      autofocus
      placeholder="new todo..."
      aria-label="Add a new Todo"
      data-id-test="input"
      @keypress.enter="addNewTodo"
    />
  </div>
</template>
<script>
import { mapActions } from "vuex";

export default {
  name: "InputTodo",
  data() {
    return {
      name: "",
    };
  },
  methods: {
    ...mapActions("todos", ["addTodo"]),
    async addNewTodo() {
      if (this.name) {
        try {
          await this.addTodo(this.name);
          this.name = "";
        } catch (error) {
          console.error(error);
        }
      }
    },
  },
};
</script>

<style scoped lang="scss">
.new-todo {
  position: relative;
  margin: 0;
  width: 100%;
  font-size: 24px;
  font-family: inherit;
  font-weight: inherit;
  line-height: 1.4em;
  border: 0;
  color: inherit;
  padding: 6px;
  border: 1px solid #999;
  box-shadow: inset 0 -1px 5px 0 rgba(0, 0, 0, 0.2);
  box-sizing: border-box;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.new-todo {
  padding: 16px 16px 16px 60px;
  border: none;
  background: rgba(0, 0, 0, 0.003);
  box-shadow: inset 0 -2px 1px rgba(0, 0, 0, 0.03);
}
</style>
