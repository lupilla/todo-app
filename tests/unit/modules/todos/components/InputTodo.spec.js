import { shallowMount } from "@vue/test-utils";
import InputTodo from "@/modules/todos/components/InputTodo.vue";
import createVuexStore from "../../../mocks/mock-store.js";
const store = createVuexStore();

describe("InputTodo tests", () => {
  let wrapper;
  let addNewTodo;

  beforeEach(() => {
    wrapper = shallowMount(InputTodo, {
      global: {
        plugins: [store],
      },
    });
  });
  test("should match with snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("should not call 'addTodo' method when writing in input and enter is not pressed", async () => {
    const input = wrapper.find('[data-id-test="input"]');
    addNewTodo = jest.spyOn(wrapper.vm, "addNewTodo");
    await input.setValue("go for a walk");
    expect(addNewTodo).not.toHaveBeenCalled();
  });

  test("should call 'addTodo' event when writing in input and enter is pressed", async () => {
    const input = wrapper.find('[data-id-test="input"]');
    addNewTodo = jest.spyOn(wrapper.vm, "addNewTodo");
    await input.setValue("go to the library");
    await input.trigger("keypress", {
      key: "enter",
    });
    expect(addNewTodo).toHaveBeenCalled();
  });

  test("'name' property should have a value when some text is written", async () => {
    const text = "cook a cheese cake";
    const input = wrapper.find('[data-id-test="input"]');
    await input.setValue(text);
    expect(wrapper.vm.$data.name).toBe(text);
  });

  test("'name' property should have an empty value by default", async () => {
    expect(wrapper.vm.$data.name).toBe("");
  });

  test("name property should not have a value when input is empty and enter is pressed", async () => {
    const input = wrapper.find('[data-id-test="input"]');
    await input.setValue("");
    await input.trigger("keypress", {
      key: "enter",
    });
    expect(wrapper.vm.$data.name).toBe("");
  });

  test("addTodo should not have been called if name property is empty", async () => {
    const addTodoSpy = jest.spyOn(wrapper.vm, "addTodo");
    const input = wrapper.find('[data-id-test="input"]');
    await input.setValue("");
    await input.trigger("keypress", {
      key: "enter",
    });
    expect(addTodoSpy).not.toHaveBeenCalled();
  });
});
