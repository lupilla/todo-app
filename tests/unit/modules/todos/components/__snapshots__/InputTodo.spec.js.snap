// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputTodo tests should match with snapshot 1`] = `<div id="newTodo"><input type="text" class="new-todo" autofocus="" placeholder="new todo..." aria-label="Add a new Todo" data-id-test="input"></div>`;
